Frame speed monitoring

ct2034 opened the pull request:
~~- [ ] Discuss whether it is a good idea to have both, an svc and topic to communicate the OM
      IMO a topic that repeatedly sends the mode would work fine.~~ postponed
- [x] update image
- [x] update docu
~~- [ ] discuss error handling~~ postponed
- [ ] squash commits

agutenkunst committed:
picking the cherries from feature/speed_observer

ct2034 committed:
draft of spec

ct2034 committed:
clarfications with @agutenkunst

ct2034 commented:
See https://github.com/PilzDE/pilz_robots/tree/feature/speed_observer for @agutenkunst 's research

ct2034 committed:
publishing a dummy message

ct2034 commented:
I think we have to implement `lookupTwist` for tf2 https://answers.ros.org/question/282492/tf-vs-tf2-lookuptwist-vs-lookup_transform/?answer=282513#post-id-282513

ct2034 committed:
building speed from velocity vector

ct2034 committed:
can observe multiple frames

ct2034 committed:
removing eigen after benchmark

ct2034 committed:
cleanup

ct2034 committed:
better initializations

ct2034 committed:
can read frames from urdf

ct2034 committed:
publishing stop when limit is exceeded

ct2034 committed:
manually finding speed

ct2034 committed:
successfully reading the speeds

ct2034 committed:
tf2-ify

ct2034 committed:
infos in tests are no errors

ct2034 committed:
an integrationtest that porbably is a unittest

ct2034 committed:
it is an unittest, actually

ct2034 committed:
this is an info

ct2034 committed:
for now working with prbt.xacro, TODO: make simplified testing xacro

ct2034 committed:
mock now working

ct2034 committed:
testing xacro

Commit:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

ct2034 committed:
Merge branch 'feature/frame_speed_monitoring' of https://github.com/PilzDE/pilz_robots into feature/frame_speed_monitoring

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

Commit:
Add component diagram for overall system architecture

Commit:
Add files for overall architecture diagram

Commit:
Fix error in overall system architecture diagram

ct2034 committed:
a threaded tf publisher for a unittest

ct2034 committed:
Merge branch 'feature/frame_speed_monitoring' of https://github.com/PilzDE/pilz_robots into feature/frame_speed_monitoring

ct2034 committed:
test needs more than 5s

Commit:
Fix overall architecture diagram

ct2034 committed:
a test with a mock

ct2034 committed:
a proper async test now

ct2034 committed:
adding expectations for stop topic

ct2034 committed:
ignoring the generated header for coverage

ct2034 committed:
merge melodic-devel to get rid of clang-tidy warning

Commit:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

Commit:
Refactor STO interface in speed_observer

ct2034 committed:
splitting up in two tests

ct2034 committed:
clang-format

ct2034 committed:
ending faster

ct2034 committed:
ending faster II

Commit:
Improve readability of speed_observer

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

ct2034 committed:
clang-formatting

ct2034 committed:
testing the timeout

Commit:
Add SetSpeedLimit srv definition

ct2034 committed:
clang-format

ct2034 committed:
correct service

ct2034 committed:
a service to set the speed limit

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

ct2034 committed:
testing the set_limit feature

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

ct2034 committed:
coverage

Commit:
Introduce time stamp to ROS operation mode message

Commit:
Remove lazy initalization from adapter_operation_mode

Commit:
Introduce operation_mode publisher to adapter_operation_mode

ct2034 committed:
testing for the actual index is stricter

ct2034 committed:
formatting

Commit:
Add operation_mode_setup_executor node

Commit:
Fix missing time_stamp update in operation_mode_setup_executor

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

ct2034 committed:
beginning of a integration test

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

ct2034 committed:
renaming

ct2034 committed:
mocks

ct2034 committed:
formatting

ct2034 committed:
fake frames to observe

ct2034 commented:
Übergabe: ich dachte an einen testcase pro Om. Man müsste dafür Dan jeweils die joint states und die additional frames mit entsprechenden Geschwindigkeit bewegen. Dann jeweils schauen, ob der stop1 getriggert wird. 

Zur Zeit ist irgendwie das Problem, dass der observer die tf nicht richtig listenen kann, obwohl mit tf_echo alles ok ist. Aber das seht ihr ja. Ich hab leider keine so richtige Idee, wo das Problem sein könnte. 

Zum mocken der geschw in joint states müsste alles soweit passen. Zum tf speed mocken, könnte ihr ggf die funktion aus dem zugehörigen unittest nehmen.

ct2034 committed:
a working xacro model and spinner in test

ct2034 commented:
> * [ ]  Discuss whether it is a good idea to have both, an svc and topic to communicate the OM
>   IMO a topic that repeatedly sends the mode would work fine.

Did you know about latched publishers (https://wiki.ros.org/roscpp/Overview/Publishers%20and%20Subscribers#Publisher_Options) this might be the thing for us!

martiniil commented:
> 
> 
> > * [ ]   Discuss whether it is a good idea to have both, an svc and topic to communicate the OM
> >   IMO a topic that repeatedly sends the mode would work fine.
> 
> Did you know about latched publishers (https://wiki.ros.org/roscpp/Overview/Publishers%20and%20Subscribers#Publisher_Options) this might be the thing for us!

Ok. I think we should have either topic or srv. Imho in this case it should be a topic.

ct2034 committed:
testing T1

ct2034 committed:
merge melodic-devel

ct2034 committed:
also testing auto

ct2034 committed:
missing from merge

ct2034 committed:
one test case only

ct2034 committed:
no need for this rather dirty workaround, then

ct2034 committed:
coverage of error msg

ct2034 committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

martiniil committed:
Add unittest for OperationModeSetupExecutor

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/get_operation_mode_func_decl.h:
```diff
@@ -0,0 +1,31 @@
+namespace prbt_hardware_support
+{
+
+using TGetOpMode = std::function<OperationModes()>;
```
Please remove the Prefix

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_operation_mode.h:
```diff
@@ -41,30 +41,25 @@ class AdapterOperationMode
    * service, the first time the function is called.
    */
-  void updateOperationMode(const int8_t new_op_mode);
+  void updateOperationMode(const OperationModes& mode);
```
Please consider postponing such changes into a separate PR in the future!

hslusarek replied:
Understood.

agutenkunst commented on prbt_hardware_support/.gitignore:
```diff
@@ -1 +1,2 @@
-doc/
+doc/html
+manifest.yaml
```
Why is this suddenly needed?

hslusarek replied:
Before the complete `doc` folder was excluded. However, I added a diagram which we want to check in. This,  in turn, is not possible if the complete `doc` folder is excluded. Regarding the `manifest.yaml` file I cannot say much except, I simply kept the status quo by still excluding it. ->Done.

hslusarek replied:
This allows us to execute the rosdoc_lite program in `prbt_hardware_support` and not have the result files under version control. -> Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/modbus_adapter_operation_mode.h:
```diff
@@ -55,6 +55,10 @@ class ModbusAdapterOperationMode : public AdapterOperationMode
   void modbusMsgCallback(const ModbusMsgInStampedConstPtr& msg_raw);
 
+private:
+  static OperationModes createOperationMode(const ModbusMsgOperationModeWrapper& msg);
```
Have you considered a member function of ModbusMsgOperationModeWrapper for getting the OperationMode?

hslusarek replied:
Good idea. Changed it accordingly. -> Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/modbus_adapter_operation_mode.h:
```diff
@@ -55,6 +55,10 @@ class ModbusAdapterOperationMode : public AdapterOperationMode
 
+private:
+  static OperationModes createOperationMode(const ModbusMsgOperationModeWrapper& msg);
+  static OperationModes createUnknownOperationMode();
```
In usage
```
updateOperationMode(createUnknownOperationMode());
```
seems more complex than
```
updateOperationMode(OperationModes::UNKNOWN);
```

hslusarek replied:
Yes. However, it is necessary to ensure that the receiver (in that case the `operation_mode_setup_executor`) can determine if this operation mode change is already known to him or not. This, in turn, is needed to ensure a proper startup phase. -> Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/operation_mode_setup_executor.h:
```diff
@@ -0,0 +1,60 @@
+namespace prbt_hardware_support
+{
+
+class OperationModeSetupExecutor
```
Please add documentation (Responsibility, Interaction, etc...)

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/operation_mode_setup_executor.h:
```diff
@@ -0,0 +1,60 @@
+  const double speed_limit_auto_;
+
+  //! Function used to propagate speed limit changes into the system.
+  TSetSpeedLimit set_speed_limit_func_;
```
Please remove the Prefix

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/set_speed_limit_func_decl.h:
```diff
@@ -0,0 +1,29 @@
+namespace prbt_hardware_support
+{
+
+using TSetSpeedLimit = std::function<bool(const double&)>;
```
Please remove the prefix

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/operation_mode_setup_executor.h:
```diff
@@ -0,0 +1,60 @@
+class OperationModeSetupExecutor
+{
+public:
+  OperationModeSetupExecutor(const double& speed_limit_t1,
```
Missing comment

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/speed_observer.h:
```diff
@@ -0,0 +1,96 @@
+
+private:
+  void triggerStop1();
+  FrameSpeeds
```
Fits on one line (120 chars)

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/speed_observer.h:
```diff
@@ -0,0 +1,96 @@
+private:
+  void triggerStop1();
+  FrameSpeeds
+  createFrameSpeedsMessage(const std::vector<double> &speeds_) const;
```
```suggestion
  createFrameSpeedsMessage(const std::vector<double> &speeds) const;
```

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/speed_observer.h:
```diff
@@ -0,0 +1,96 @@
+
+private:
+  ros::NodeHandle nh_;
+  const std::string reference_frame_;
```
Missing comments

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/speed_observer.h:
```diff
@@ -0,0 +1,96 @@
+  bool isWithinLimit(const double &speed) const;
+
+private:
+  static double speedFromTwoPoses(const tf2::Vector3 &a, const tf2::Vector3 &b,
```
Fits into one line

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/speed_observer.h:
```diff
@@ -0,0 +1,96 @@
+  return d / t;
+}
+
+inline bool SpeedObserver::isWithinLimit(const double &speed) const {
```
```suggestion
inline bool SpeedObserver::isWithinLimit(const double &speed) const 
{
```

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/speed_observer.h:
```diff
@@ -0,0 +1,96 @@
+}
+
+// This method is only needed for the unittest
+inline void SpeedObserver::terminateNow() {
```
```suggestion
inline void SpeedObserver::terminateNow() 
{
```

agutenkunst commented on prbt_hardware_support/launch/operation_mode_setup_executor_node.launch:
```diff
@@ -0,0 +1,27 @@
+
+<launch>
+
+  <arg name="speed_limit_t1" default="250" />
```
ROS default are SI units. If you are not using them please make explicit in the variable name

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/launch/operation_mode_setup_executor_node.launch:
```diff
@@ -0,0 +1,27 @@
+<launch>
+
+  <arg name="speed_limit_t1" default="250" />
+  <arg name="speed_limit_automatic" default="5000" />
```
Same as above

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/msg/FrameSpeeds.msg:
```diff
@@ -0,0 +1,20 @@
+Header header
+
+string[] name
+float64[] speed
```
m/s correct?

martiniil committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

agutenkunst commented on prbt_hardware_support/spec.dox:
```diff
@@ -123,4 +123,21 @@
+ *  - (New) The user can define additional tf frames to be monitored.
+ *
+ * @definereq{Stop1_on_violation_of_speed_limit}
+ *  - (New) When one of the montired speeds exceeds 250mms^-1
```
```suggestion
 *  - (New) When one of the monitored speeds exceeds 250mms^-1
```

agutenkunst replied:
Not entirely correct. It depends on the OperationMode

agutenkunst commented on prbt_hardware_support/src/adapter_operation_mode.cpp:
```diff
@@ -21,45 +21,42 @@ namespace prbt_hardware_support
 static const std::string SERVICE_NAME_GET_OPERATION_MODE = "/prbt/get_operation_mode";
+static const std::string TOPIC_OPERATION_MODE = "/prbt/operation_mode";
+
+static constexpr int DEFAULT_QUEUE_SIZE_MODBUS{10};
```
Variable name?

ct2034 replied:
?!?!

hslusarek replied:
@agutenkunst Yea, whats wrong with the variable name? It states it purpose.

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/speed_observer.h:
```diff
@@ -0,0 +1,96 @@
+                                               const tf2::Vector3 &b,
+                                               const double &t) {
+  double d = tf2::tf2Distance(a, b);
+  return d / t;
```
Please handle the div by zero!

agutenkunst commented on prbt_hardware_support/src/operation_mode_setup_executor.cpp:
```diff
@@ -0,0 +1,73 @@
+    break;
+  }
+
+  if (set_speed_limit_func_)
```
If not?

martiniil replied:
I would suggest to move the discussion about handling of error cases out of this pr.
@agutenkunst ok?

agutenkunst commented on prbt_hardware_support/src/operation_mode_setup_executor_node.cpp:
```diff
@@ -0,0 +1,109 @@
+
+static constexpr uint32_t DEFAULT_QUEUE_SIZE {10} ;
+
+// LCOV_EXCL_START
```
```suggestion
```

hslusarek replied:
Done.

agutenkunst commented on prbt_hardware_support/src/operation_mode_setup_executor_node.cpp:
```diff
@@ -0,0 +1,109 @@
+  SetSpeedLimit srv_msg;
+  srv_msg.request.speed_limit = speed_limit;
+  bool call_success = srv_client.call(srv_msg);
+  if (!call_success)
```
In this case the speed limit could not be set? Should this be handled?

martiniil replied:
outdated. See above conversation.

agutenkunst commented on prbt_hardware_support/src/speed_observer.cpp:
```diff
@@ -0,0 +1,185 @@
+  ros::Time previous_t = now;
+
+  // Starting the observer loop
+  while (ros::ok() && !terminate_)
```
maybe construct a function ```observationLoop``` ?

ct2034 replied:
why?

martiniil replied:
I would keep it like it is.

agutenkunst commented on prbt_hardware_support/src/speed_observer.cpp:
```diff
@@ -0,0 +1,185 @@
+      }
+      try
+      {
+        waitTillTFReady(frame, now);
```
With `WAITING_TIME_FOR_TRANSFORM_S` beeing 1 second and 10 retries you could end up with a long time until every frame is checked. Please consider your worst cases...

ct2034 replied:
This is really only relevant if there is a problem with TF to begin with. For example if one of the frames is missing. In normal operation, you will never wait for more than 1 Cycle on the first frame. That is also why you only get a warning after the first wait. And I think a case where TFs come available one after the other, for the same query time is very unlikely.

martiniil committed:
Test handling of operation mode UNKNOWN

agutenkunst commented on prbt_hardware_support/src/speed_observer.cpp:
```diff
@@ -0,0 +1,185 @@
+  bool call_success = sto_client_.call(sto_srv);
+  if (!call_success)
+  {
+    ROS_ERROR_STREAM("No success calling service: " << sto_client_.getService());
```
This could even be a case for ROS_FATAL_STREAM should this observer handle. What should happen in this case?

ct2034 replied:
Yes, maybe we should shut down the node then and make it required. What do you think? And about the severity: Idk, have you ever used fatal before?

martiniil replied:
See above

> 
> 
> I would suggest to move the discussion about handling of error cases out of this pr.
> @agutenkunst ok?

agutenkunst commented on prbt_hardware_support/src/speed_observer.cpp:
```diff
@@ -0,0 +1,185 @@
+  }
+  else if (!sto_srv.response.success)
+  {
+    ROS_ERROR_STREAM("Service: " << sto_client_.getService() << " failed with error message:\n"
```
See above

martiniil replied:
See above

> 
> 
> I would suggest to move the discussion about handling of error cases out of this pr.
> @agutenkunst ok?

agutenkunst commented on prbt_hardware_support/src/speed_observer_node.cpp:
```diff
@@ -0,0 +1,73 @@
+  ros::NodeHandle pnh{ "~" };
+  ros::NodeHandle nh;
+
+  urdf::Model m;
```
naming!

agutenkunst commented on prbt_hardware_support/src/speed_observer_node.cpp:
```diff
@@ -0,0 +1,73 @@
+  std::vector<urdf::LinkSharedPtr> links;
+  std::vector<std::string> frames_to_observe;
+  m.getLinks(links);
+  ROS_DEBUG_STREAM("frames_to_observe (from urdf):");
```
```suggestion
  ROS_DEBUG_STREAM("Received the following frames to observer from urdf: ");
```
The output is for the user not for the programmer!

ct2034 replied:
Its a debug msg ...

agutenkunst commented on prbt_hardware_support/src/speed_observer_node.cpp:
```diff
@@ -0,0 +1,73 @@
+  std::vector<std::string> frames_to_observe;
+  m.getLinks(links);
+  ROS_DEBUG_STREAM("frames_to_observe (from urdf):");
+  for (const auto& l : links)
```
```suggestion
  for (const auto& link : links)
```

agutenkunst replied:
Also std::transform should work here...

agutenkunst commented on prbt_hardware_support/src/speed_observer_node.cpp:
```diff
@@ -0,0 +1,73 @@
+  }
+  std::vector<std::string> additional_frames;
+  pnh.getParam(ADDITIONAL_FRAMES_PARAM_NAME, additional_frames);
+  ROS_DEBUG_STREAM("frames_to_observe (additional_frames): ");
```
```suggestion
  ROS_DEBUG_STREAM("Additional frames defined by user: ");
```

agutenkunst commented on prbt_hardware_support/src/speed_observer_node.cpp:
```diff
@@ -0,0 +1,73 @@
+  std::vector<std::string> additional_frames;
+  pnh.getParam(ADDITIONAL_FRAMES_PARAM_NAME, additional_frames);
+  ROS_DEBUG_STREAM("frames_to_observe (additional_frames): ");
+  for (const auto& f : additional_frames)
```
```suggestion
  for (const auto& frame : additional_frames)
```

agutenkunst commented on prbt_hardware_support/src/speed_observer_node.cpp:
```diff
@@ -0,0 +1,73 @@
+      nh.advertiseService(SET_SPEED_LIMIT_SERVICE, &SpeedObserver::setSpeedLimitCb, &observer);
+  observer.startObserving(OBSERVATION_FREQUENCY);
+
+  return EXIT_SUCCESS;
```
Consider combining this with a return value of startObserving to return EXIT_FAILURE in special cases

ct2034 replied:
in what cases? why? what for?

martiniil replied:
See above

> 
> 
> I would suggest to move the discussion about handling of error cases out of this pr.
> @agutenkunst ok?

agutenkunst commented on prbt_hardware_support/srv/SetSpeedLimit.srv:
```diff
@@ -0,0 +1,18 @@
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+float64 speed_limit
```
Unit?

ct2034 replied:
m/s

agutenkunst commented on prbt_hardware_support/test/config/speed_observer_test.xacro:
```diff
@@ -0,0 +1,42 @@
+  <link name="world"/>
+
+  <joint name="$(arg robot_prefix)world-a" type="continuous">
+  <origin xyz="0 0 1" rpy="0 1.57079632679 0" />
```
```suggestion
  <origin xyz="0 0 1" rpy="0 ${pi/2} 0" />
```

agutenkunst commented on prbt_hardware_support/test/config/speed_observer_test.xacro:
```diff
@@ -0,0 +1,42 @@
+  <origin xyz="0 0 1" rpy="0 1.57079632679 0" />
+    <parent link="world"/>
+    <child link="$(arg robot_prefix)a"/>
+    <limit effort="999" velocity="99" />
```
```suggestion
```
Needed?

ct2034 replied:
yes

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+  nh_.shutdown();
+}
+
+void SpeedObserverIntegrationTest::publishOm(OperationModes::_value_type omv){
```
```suggestion
void SpeedObserverIntegrationTest::publishOperationMode(OperationModes::_value_type omv){
```

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+ *
+ * Expected Results:
+ *    0. -
+ *    1. Operation mode not to be called any more
```
```suggestion
 *    1. Operation mode service callback not to be called any more
```

martiniil replied:
Outdated

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+  EXPECT_CALL(*this, sto_cb(StoState(true), _))
+      .WillRepeatedly(DoAll(SetArgReferee<1>(sto_res), ACTION_OPEN_BARRIER(BARRIER_STOP_HAPPENED)));
+
+  std::thread pubisher_thread = std::thread(&SpeedObserverIntegrationTest::publishJointStatesAtSpeed, this, 1.0);
```
How can u be sure that the Limit was triggered at the configured frame speed?

ct2034 replied:
what do you mean?

martiniil replied:
we cannot be sure.

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+ *    4. Sto is triggered
+ *    5. Sto is triggered and error message is produced
+ */
+TEST_F(SpeedObserverIntegrationTest, testT1)
```
Correct Test Name? I am missing the test for the additional frame...

martiniil replied:
Done

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_operation_mode_setup_executor.cpp:
```diff
@@ -0,0 +1,230 @@
+ *  1. Call updateOperationMode() with operation mode UNKNOWN and current time stamp.
+ *
+ * Expected Results:
+ *  1. setSpeedLimit() is called with limit less or equal 0.0.
```
Please add spec and ref

martiniil replied:
Done.

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+MATCHER_P(ContainsAllNames, names,
+          "Names" + PrintToString(names) + std::string(negation ? "are not" : "are") + " in message.")
+{
+  for (auto& n : names)
```
```suggestion
  for (auto& name : names)
```

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+    for (auto& argn : arg->name)
+      if (argn.compare(n) == 0)
+        found = true;
+    if (!found)
```
Braces

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+  {
+    bool found{ false };
+    for (auto& argn : arg->name)
+      if (argn.compare(n) == 0)
```
Braces

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+  for (auto& n : names)
+  {
+    bool found{ false };
+    for (auto& argn : arg->name)
```
Braces

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+
+  pnh_.getParam(ADDITIONAL_FRAMES_PARAM_NAME, additional_frames_);
+  ROS_DEBUG_STREAM("SetUp pnh:" << pnh_.getNamespace() << " nh:" << nh_.getNamespace());
+  for (auto& f : additional_frames_)
```
```suggestion
  for (auto& frame : additional_frames_)
```

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+  operation_mode_pub_.publish(om);
+}
+
+void SpeedObserverIntegrationTest::publishJointStatesAtSpeed(double v)
```
```suggestion
void SpeedObserverIntegrationTest::publishJointStatesAtSpeed(double vel)
```

ct2034 replied:
really?

ct2034 replied:
is vel better than v? Why not velocity, then?

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp:
```diff
@@ -0,0 +1,290 @@
+ *    1. Operation mode not to be called any more
+ *       Sto is triggered
+ *    2. -
+ *    3. Operation mode not to be called any more
```
Service?

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.test:
```diff
@@ -0,0 +1,60 @@
+
+    <node pkg="tf" type="static_transform_publisher"
+        name="test_transform_publisher"
+        args="1 1 0 0 1.57 0 /world /$(arg testing_frame) 100"/>
```
Missing tests for the additional frame

martiniil replied:
Done.

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_speed_observer.cpp:
```diff
@@ -0,0 +1,406 @@
+    ros::Time current = ros::Time::now();
+    t = (current - start).toSec();
+    // a has no speed
+    geometry_msgs::TransformStamped tfsa;
```
naming

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_speed_observer.cpp:
```diff
@@ -0,0 +1,406 @@
+    tfsa.transform.translation.z = 99.0;
+    tfsa.transform.rotation.w = 1;
+    // b has speed v
+    geometry_msgs::TransformStamped tfsb;
```
naming

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_speed_observer.cpp:
```diff
@@ -0,0 +1,406 @@
+ *
+ * Test Sequence:
+ *    0. Starting up
+ *    1. Publishing tf movements that *not* have a too high speed
```
```suggestion
 *    1. Publishing tf movements that are withing the speed limit
```

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_speed_observer.cpp:
```diff
@@ -0,0 +1,406 @@
+ *
+ * Expected Results:
+ *    0. -
+ *    1. Correct values are publshed on the speed topic.
```
```suggestion
 *    1. Correct values are published on the speed topic.
```

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_speed_observer.cpp:
```diff
@@ -0,0 +1,406 @@
+void SpeedObserverUnitTest::publishTfAtSpeed(double v)
+{
+  static tf2_ros::TransformBroadcaster br;
+  ros::Rate r = ros::Rate(TEST_FREQUENCY * 3);  // publishing definitely faster then observing
```
If you increase the frequence does the noise reduce? I see you have to cover quite a range in your expectations

ct2034 replied:
no, not in my experience

agutenkunst reviewed (changes requested):
I will recheck the tests again. But so far I also found some issues.

Also missing:
- [ ] Integration into the robot launch file?
- [ ] Check on the real system

agutenkunst commented on prbt_hardware_support/src/speed_observer.cpp:
```diff
@@ -0,0 +1,185 @@
+  : nh_(nh), reference_frame_(reference_frame), frames_to_observe_(frames_to_observe)
+{
+  frame_speeds_pub_ = nh.advertise<FrameSpeeds>(FRAME_SPEEDS_TOPIC_NAME, DEFAULT_QUEUE_SIZE);
+  waitForService(STO_SERVICE);
```
In your image this is a Topic...

ct2034 committed:
documentation

ct2034 committed:
starting to make operation mode topic only

ct2034 committed:
Merge branch 'feature/frame_speed_monitoring' of https://github.com/PilzDE/pilz_robots into feature/frame_speed_monitoring

ct2034 committed:
formatting

martiniil committed:
Update architecture diag: STO service

ct2034 committed:
Merge branch 'feature/frame_speed_monitoring' of https://github.com/PilzDE/pilz_robots into feature/frame_speed_monitoring

ct2034 committed:
spec clarifications

ct2034 committed:
handling div by zero

ct2034 committed:
incorpsorating review in node

ct2034 committed:
Update prbt_hardware_support/test/config/speed_observer_test.xacro

Co-Authored-By: Alexander Gutenkunst <a.gutenkunst@pilz.de>

ct2034 committed:
Update prbt_hardware_support/test/unittests/unittest_speed_observer.cpp

Co-Authored-By: Alexander Gutenkunst <a.gutenkunst@pilz.de>

ct2034 committed:
Update prbt_hardware_support/test/unittests/unittest_speed_observer.cpp

Co-Authored-By: Alexander Gutenkunst <a.gutenkunst@pilz.de>

ct2034 committed:
naming and braces

agutenkunst reviewed (changes requested):
Please update the image, some service names are wrong

Commit:
Incorporation of review comments

* Removing previx "T" from function type declaration
* Adding code docu to OperationModeSetup exeutor
* Renaming const variable name in AdapterOperationMode

Commit:
Change speed limits in launch file to SI units

Commit:
Update specification and add linking to unittests of OperationModeSetupExecutor

Commit:
Fix doxygen problems

Commit:
Update gitignore file

Commit:
Revert "Update gitignore file"

This reverts commit 0394ede386c0fda20d8606e9c07506b32e49d6bd.

Commit:
Rename overview md file in doc folder

ct2034 committed:
variable naming

ct2034 committed:
integration testing addtioninal tf speeds

ct2034 committed:
Merge branch 'feature/frame_speed_monitoring' of https://github.com/PilzDE/pilz_robots into feature/frame_speed_monitoring

Commit:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

Commit:
Add tests for service calls in operation_mode_setup_executor_node

Commit:
Fix clang tidy errors in integrationtest_speed_observer

Commit:
Exclude FrameSpeeds.msg from coverage

ct2034 committed:
improving test / spec documentation

Commit:
Exclude SetSpeedLimit service from coverage

ct2034 committed:
formatting

ct2034 committed:
Merge branch 'feature/frame_speed_monitoring' of https://github.com/PilzDE/pilz_robots into feature/frame_speed_monitoring

ct2034 committed:
clang-tidy issues

ct2034 committed:
merge feature/frame_speed_monitoring

ct2034 committed:
removing function calls and adding topic publishs

ct2034 commented:
the removing of the service should be almost finished. Just the `integrationtest_operation_mode` needs some adoption as you will be able to tell if you compile it ;-)

martiniil commented on prbt_hardware_support/src/operation_mode_setup_executor.cpp:
```diff
@@ -23,20 +23,15 @@ namespace prbt_hardware_support
-    ROS_DEBUG("Operation Mode at init: %d", om.value);
-    updateOperationMode(om);
+    ROS_WARN("Waiting for first operation mode");
+    ros::Duration(1).sleep();
```
@ct2034 This is an endless loop, the construction of the executor never finishes.

martiniil committed:
Remove superfluous (infinite) wait loop

martiniil committed:
Update unittest_operation_mode_setup_executor

martiniil commented:
>     * [ ]  Discuss whether it is a good idea to have both, an svc and topic to communicate the OM
>       IMO a topic that repeatedly sends the mode would work fine.

https://wiki.ros.org/ROS/Patterns/Communication#Communication_via_Topics_vs_Services_vs_X

Since we only act on changes and do not have a continuous data stream, a service should be preferred. To be more consistent and also follow the above patterns, we should get rid of the topic.

Also we should be aware of https://github.com/ros/ros_comm/issues/286#issuecomment-25334904:

> What you describe sounds like a common misuse of pub/sub - assuming that all messages in the start-up phase will be delivered.

martiniil committed:
Fix adapter_operation_mode

With unknown initial operation mode the adapter would not publish a message
because there was no operation mode switch.

martiniil committed:
Refactor integrationtest_operation_mode

* service GetOperationMode no longer exists
* Use AsyncTest barriers instead of sleep

martiniil committed:
Remove GetOperationMode service artefacts

martiniil committed:
Make service client non-persistent

martiniil committed:
Fix tests

martiniil committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

martiniil committed:
Restore GetOperationMode service

In order to not brake existing code in pilz_industrial_motion.

martiniil committed:
Restore line coverage

martiniil committed:
Stabilize integrationtest of speed observer

martiniil committed:
Restore frequency in test

martiniil committed:
Integrate operation mode setup into launch process

martiniil committed:
Make speed observer required

martiniil committed:
Remove unused constants

martiniil committed:
Merge branch 'melodic-devel' into feature/frame_speed_monitoring

martiniil commented:
the whole frame-speed-computing needs way too much time:

> [ INFO] [1568895417.853714392]: Observing at 50.0Hz
[ERROR] [1568895418.855039500]: DISPLAY ACTUAL CYCLE TIME: 1.001590906

> [1568896381.342707520]: Waited 1.00 s for transform of frame prbt_flange

I found the problem: this constant has to be chosen small enough
`  static constexpr double WAITING_TIME_FOR_TRANSFORM_S{ 0.1 };`

martiniil committed:
Do not wait that long during transform lookups

martiniil committed:
Fix launch file

martiniil committed:
Fix sto service call

martiniil committed:
Update docu

martiniil committed:
Fix integrationtest

martiniil committed:
Increase configurability of integrationtest

martiniil committed:
Update architecture diagram

martiniil committed:
Update changelog

martiniil committed:
Fix doxygen warning

Commit:
Update OperationModeSetupExecutorNode docu

martiniil committed:
Update architecture image

martiniil committed:
Fix architecture.png

martiniil committed:
Remove waiting in while loop of speed_observer

martiniil committed:
Remove superfluous waiting for latest transform

martiniil committed:
Fix topic string

martiniil committed:
refactor speed_observer
* exclude frames which cannot move
* remove waiting in observation-loop
* always latest tf transforms and their time stamps

martiniil committed:
Increase observer frequency

martiniil commented:
I tested with a full-speed PTP. The observer runs at 20Hz frequency. The output shows a speed of 0.41 m/s. I think in order to avoid this, we have to adopt preventive measures.

> [ INFO] [1569329475.516449984]: Generating PTP trajectory...
[ERROR] [1569329475.828727190]: Speed 0.25 m/s of frame >prbt_flange< exceeds limit of 0.25 m/s
[ERROR] [1569329475.831057345]: Speed 0.26 m/s of frame >prbt_link_5< exceeds limit of 0.25 m/s
[ INFO] [1569329475.832972228]: Controller prbt/manipulator_joint_trajectory_controller/ successfully finished
[ERROR] [1569329475.832974032]: Speed 0.25 m/s of frame >prbt_tcp< exceeds limit of 0.25 m/s
[ WARN] [1569329475.833042323]: Controller handle prbt/manipulator_joint_trajectory_controller/ reports status PREEMPTED
[ INFO] [1569329475.833078280]: Completed trajectory execution with status PREEMPTED ...
[ERROR] [1569329475.834638004]: Speed 0.25 m/s of frame >prbt_tool0< exceeds limit of 0.25 m/s
[ERROR] [1569329475.878745913]: Speed 0.30 m/s of frame >prbt_flange< exceeds limit of 0.25 m/s
[ERROR] [1569329475.881019365]: Speed 0.31 m/s of frame >prbt_link_5< exceeds limit of 0.25 m/s
[ERROR] [1569329475.883136407]: Speed 0.30 m/s of frame >prbt_tcp< exceeds limit of 0.25 m/s
[ERROR] [1569329475.884836507]: Speed 0.35 m/s of frame >prbt_tool0< exceeds limit of 0.25 m/s
[ERROR] [1569329475.928735866]: Speed 0.39 m/s of frame >prbt_flange< exceeds limit of 0.25 m/s
[ERROR] [1569329475.930819360]: Speed 0.39 m/s of frame >prbt_link_5< exceeds limit of 0.25 m/s
[ERROR] [1569329475.933469383]: Speed 0.39 m/s of frame >prbt_tcp< exceeds limit of 0.25 m/s
[ERROR] [1569329475.935460360]: Speed 0.41 m/s of frame >prbt_tool0< exceeds limit of 0.25 m/s
[ERROR] [1569329475.978723533]: Speed 0.41 m/s of frame >prbt_flange< exceeds limit of 0.25 m/s
[ERROR] [1569329475.980536406]: Speed 0.38 m/s of frame >prbt_link_5< exceeds limit of 0.25 m/s
[ERROR] [1569329475.982269155]: Speed 0.39 m/s of frame >prbt_tcp< exceeds limit of 0.25 m/s
[ERROR] [1569329475.983961341]: Speed 0.39 m/s of frame >prbt_tool0< exceeds limit of 0.25 m/s
[ERROR] [1569329476.028735881]: Speed 0.34 m/s of frame >prbt_flange< exceeds limit of 0.25 m/s
[ERROR] [1569329476.030658719]: Speed 0.29 m/s of frame >prbt_link_5< exceeds limit of 0.25 m/s
[ERROR] [1569329476.032528609]: Speed 0.30 m/s of frame >prbt_tcp< exceeds limit of 0.25 m/s
[ INFO] [1569329476.034993333]: Halting down XXX
[ERROR] [1569329476.037497240]: Speed 0.30 m/s of frame >prbt_tool0< exceeds limit of 0.25 m/s
[ INFO] [1569329482.579385039]: Recovering XXX

martiniil commented:
I used rate 100 for the joint_state_controller as well as the robot_state_publisher, resulting in avg rate 65 of /tf with maximal time 0.03s (=33 hz)

> subscribed to [/prbt/joint_states]
average rate: 99.864
        min: 0.006s max: 0.016s std dev: 0.00093s window: 95

> subscribed to [/joint_states]
average rate: 100.009
        min: 0.008s max: 0.012s std dev: 0.00038s window: 95

> subscribed to [/tf]
average rate: 65.458
        min: 0.008s max: 0.030s std dev: 0.00515s window: 62

martiniil committed:
Introduce launch arg frequency

martiniil committed:
Update docu, include units of parameters

Commit:
Fix unused parameter in speed_observer class

Commit:
Fix testTimeout test in speed observer unit test

martiniil committed:
Insert linebreaks

martiniil committed:
Fix speed observer unit test
* remove test for observing at a rate >> tf

martiniil committed:
Fix speed observer integration test

Commit:
Fix line coverage in SpeedObserver

Commit:
Fix spelling

Commit:
Exclude parameter reading from line coverage in speed_observer_node

martiniil committed:
Add missing docu of parameter

martiniil committed:
Fix sentence in spec.dox

martiniil committed:
Fix lists in test docus for doxygen

martiniil committed:
Operation mode T2 is not supported yet
